Add epics-base for linux only

ericdill opened the pull request:
This might take a while. It's a tricky one to build.  cc @klauer @tacaswell

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/epics-base```) and found some lint.

Here's what I've got...


For **recipes/epics-base**:

 * The recipe could do with some maintainers listed in the "extra/recipe-maintainers" section.
 * The recipe must have some tests.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/epics-base```) and found some lint.

Here's what I've got...


For **recipes/epics-base**:

 * The recipe must have some tests.

jakirkham commented on recipes/epics-base/meta.yaml:
```diff
@@ -0,0 +1,36 @@
+build:
+  number: 0
+  skip: True  # [win]
+  skip: True  # [osx]
```
Probably should combine them to be `skip: true  # [not linux]`.

ericdill commented:
I can get this to build in the linux-anvil dockerfile locally if I install readline-devel and perl-devel. And I just figured out that `yum_requirements.txt` is a thing, so yay!  

Is this a valid option?  I mean I would imagine it is fine since these are build requirements not run requirements.  I am a little concerned that the `conda-forge/readline` package does not seem to be working though. Or maybe the readline conda package is not meant for linking against?  The error I'm seeing is:

```
/usr/bin/g++ -o libCom.so.3.14 -shared -fPIC -Wl,-hlibCom.so.3.14         -m64                bucketLib.o epicsRingPointer.o epicsRingBytes.o postfix.o calcPerform.o cvtFast.o resourceLib.o epicsOnce.o epicsSingletonMutex.o dbmf.o ellLib.o envSubr.o envData.o errlog.o errSymLib.o errSymTbl.o fdmgr.o fdManager.o freeListLib.o gpHashLib.o iocsh.o registry.o libComRegister.o iocLog.o logClient.o macCore.o macEnv.o macUtil.o aToIPAddr.o adjustment.o cantProceed.o epicsConvert.o epicsExit.o epicsStdlib.o epicsString.o truncateFile.o ipAddrToAsciiAsynchronous.o epicsUnitTest.o epicsThread.o epicsMutex.o epicsEvent.o epicsTime.o epicsMessageQueue.o epicsMath.o epicsGeneralTime.o osiClockTime.o osdSock.o osdSockAddrReuse.o osiSock.o systemCallIntMech.o epicsSocketConvertErrnoToString.o osdAssert.o osdFindSymbol.o osdInterrupt.o osdPoolStatus.o osdSignal.o osdEnv.o epicsReadline.o epicsTempFile.o epicsStdio.o osdStdio.o osdThread.o osdMutex.o osdEvent.o osdTime.o osdProcess.o osdNetIntf.o osdMessageQueue.o devLibVME.o devLibVMEOSD.o taskwd.o epicsTimer.o timer.o timerQueue.o timerQueueActive.o timerQueueActiveMgr.o timerQueuePassive.o tsDefs.o      -lpthread   -lreadline -lm -lrt -ldl -lgcc
/usr/bin/ranlib libCom.a
/opt/rh/devtoolset-2/root/usr/bin/ld: cannot find -lreadline
```

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/epics-base```) and found it was in an excellent condition.

jakirkham commented:
As those are both packages in conda-forge, I would prefer we use them. If there are issues with them then, this seems like a good time to get them sorted. Could you please switch back so we can see the build logs?

ericdill commented:
So my local build followed these steps to figure out what was wrong and get it running in the condaforge/linux-anvil docker container. I will reproduce these steps on github with circleci

1. Saw an error with readline build/run requirements. 3361089; [circleci link] (https://circleci.com/gh/ericdill/staged-recipes/31)
`/opt/rh/devtoolset-2/root/usr/bin/ld: cannot find -lreadline`

2. Installed the readline-devel package in the container and ran it with *no* build/run dependencies and saw a perl error f3d40e6; [circleci link] (https://circleci.com/gh/ericdill/staged-recipes/32):
`Can't open perl script "/usr/share/perl5/ExtUtils/xsubpp": No such file or directory`

3. Added a perl build dependency. Saw the same perl error. Also don't know a lot about perl so wasn't sure how to solve this problem. 2d7aba3; [circleci link] (https://circleci.com/gh/ericdill/staged-recipes/33)
`Can't open perl script "/usr/share/perl5/ExtUtils/xsubpp": No such file or directory`

4. Removed perl dependency and Installed perl-devel with yum in the container and ran it.   Successful build. 39d1d41; [circleci link] (https://circleci.com/gh/ericdill/staged-recipes/34)

In terms of the system packages that are available through `yum`, there is a `readline` package and a `readline-devel` package.  You need the `readline-devel` package to build this recipe, `readline` by itself wont work.  I wonder if that is the same problem as I am seeing here with the conda `readline` package.  Is it possible to make a conda package equivalent to `readline-devel`?

jakirkham commented:
Thanks for the explanation. So, the reason I am asking we have the same recipe up here using the conda-forge packages in a new commit is it will make it easier and clearer for us to see what is going.

Additionally, it will give others (myself included) a chance to play with it either locally or in a separate fork with CI enabled. That way we can make a few suggestions and possibly PR changes back. In the process, we can tweak other packages like `perl` and `readline` as needed. Right now, I'm a little unsure about what this recipe looks like exactly.

Would it be ok if it were added to a new commit here so we can see it?

ericdill commented:
I think we are not communicating effectively. Points 1-4 are the last four commits in this repository. Isn't that what you're asking for?

jakirkham commented:
> I think we are not communicating effectively. Points 1-4 are the last four commits in this repository. Isn't that what you're asking for?

That may be true. I may be too tired to really dig into this tonight. 😪

That being said, I do understand that you are trying to tease out issues with each package respectively. While that seems like a reasonable thing to do, what would really help me is just having one commit that has all of  problems in it. If that's one of these commits up here, please point me to that one sha.

ericdill commented:
> what would really help me is just having one commit that has all of problems in it.

Ah I see. You don't see the readline problem and the perl problem with the one commit. The build hits the readline problem and bails before it has a chance to fail due to lacking the right perl binary.  I'll make that change and push it up when I get back to my laptop (currently on baby duty).  Basically the requirements section should look like 

```yaml
requirements:
  build:
    - readline 
    - perl
  run:
    - readline
```
And no `yum_requirements.txt`

ericdill commented:
> I may be too tired to really dig into this tonight.

Go get some rest 😃.  This can be dealt with later

ericdill commented:
> one commit that has all of problems in it

80e4560

ericdill commented:
Ok, making progress.  Apparently I need to explicitly tell `make` to look at `$LD_RUN_PATH` for the readline lib. Is that the right place? (See change in 3e43e94)

ericdill commented:
@klauer Do you have any idea why this line:
```make
  EXTUTILS = $(shell $(PERL) ../perlConfig.pl privlib)/ExtUtils
```
in `src/cap5/Makefile` results in a value of EXTUTILS=`.../../lib/perl5/5.20.3/ExtUtils`?  I can't help but think that the three dots at the beginning is screwing something up...

You can see this error in the circleci build output: https://circleci.com/gh/ericdill/staged-recipes/36

klauer commented:
@ericdill Never saw that before. `perlConfig.pl` is very simple, though: 
```perl
use Config; 
my $arg = shift;
print $Config{$arg};
```
`Config` is a [core perl module](http://perldoc.perl.org/Config.html), not an EPICS thing. This should just print out the lib directory.

So my guess would be that there's some strangely-configured, conda-packaged perl it's getting instead of the system one?

ericdill commented:
Interesting.  I am very unfamiliar with perl, but I can at least provide steps to get to a state where this might be debuggable.

```bash
git clone https://github.com/ericdill/staged-recipes
cd staged-recipes
git checkout add-epics-base
docker run -it -v `pwd`:/repo condaforge/linux-anvil /bin/bash
conda config --add channels conda-forge
conda install conda-build-all --yes
conda-build-all /repo/recipes/epics-base
# the above command will fail with the perl issue, but will leave the source 
# extracted for you in `/opt/conda/conda-bld/work` and the build environment 
# at `/opt/conda/envs/_build`, so let's activate that environment
source activate /opt/conda/envs/_build
# and then navigate to the location of this perlConfig.pl module
cd /opt/conda/conda-bld/work/base-3.14.12.4/src/cap5
# and run the perlConfig script
echo $(perl perlConfig.pl privlib)
```
will produce:
```
.../../lib/perl5/5.20.3
```
and
```bash
ls $(perl perlConfig.pl privlib)
```
produces:
```
ls: cannot access .../../lib/perl5/5.20.3: No such file or directory
```

Interestingly, system perl seems to work just fine here. (Continuing execution of commands from previous bash block)

```bash
source deactivate
echo $(perl perlConfig.pl privlib)
```
produces:
```
/usr/share/perl5
```
and 
```bash
ls  $(perl perlConfig.pl privlib)
```
produces
```
abbrev.pl       bigint.pl       constant.pm     English.pm     filetest.pm   IO            open3.pl        Shell.pm       Text           version
AnyDBM_File.pm  bigint.pm       CORE.pod        Env.pm         Filter        IPC           open.pm         shellwords.pl  Thread         version.pm
Archive         bignum.pm       ctime.pl        Error          FindBin.pm    less.pm       overload        sigtrap.pm     Thread.pm      version.pod
assert.pl       bigrat.pl       DBM_Filter      Error.pm       finddepth.pl  Locale        overloading.pm  sort.pm        Tie            vmsish.pm
Attribute       bigrat.pm       DBM_Filter.pm   exceptions.pl  find.pl       locale.pm     overload.pm     stat.pl        Time           warnings
attributes.pm   blib.pm         DB.pm           Exporter       flush.pl      Log           perl5db.pl      strict.pm      timelocal.pl   warnings.pm
autodie         bytes_heavy.pl  Devel           Exporter.pm    getcwd.pl     look.pl       PerlIO          subs.pm        Unicode
autodie.pm      bytes.pm        diagnostics.pm  ExtUtils       Getopt        Math          PerlIO.pm       Switch.pm      unicore
AutoLoader.pm   cacheout.pl     Digest          fastcwd.pl     getopt.pl     Memoize       pod             Symbol.pm      UNIVERSAL.pm
AutoSplit.pm    Carp            Digest.pm       Fatal.pm       getopts.pl    Memoize.pm    Pod             syslog.pl      User
autouse.pm      Carp.pm         DirHandle.pm    feature.pm     hostname.pl   Module        pwd.pl          tainted.pl     utf8_heavy.pl
B               charnames.pm    dotsh.pl        fields.pm      I18N          Net           Safe.pm         Term           utf8.pm
base.pm         Class           Dumpvalue.pm    File           if.pm         newgetopt.pl  Search          termcap.pl     validate.pl
Benchmark.pm    complete.pl     dumpvar.pl      FileCache.pm   importenv.pl  NEXT.pm       SelectSaver.pm  Test           vars.pm
bigfloat.pl     Config          encoding        FileHandle.pm  integer.pm    open2.pl      SelfLoader.pm   Test.pm        vendor_perl
```

@jakirkham any ideas who might know more about conda packaged perl?

ericdill commented:
After a little more investigation it seems like there are these wonky paths with triple dots in conda packaged perl that is shipped from the `conda-forge` and `dan_blanchard` channels.  That seems like a problem with the relocation since I don't think that unix paths can have these triple dots in them.  At least I was unsuccessful in finding any information about triple dots (`...`) in unix paths via google.

```bash
conda create -p /tmp/foo perl -c conda-forge --yes
cd /tmp/foo
grep -r "\.\.\.\/"
```
[link to the output of the above command] (https://gist.github.com/ericdill/e2eba39cc6b29b74d087fb3e3d372cba) since it is sort of long

I'll cc @msarahan too since you are one of the maintainers of the Perl package.  Any idea whats up?

edit: After more googling, I seem to have stumbled on [this thread] (http://www.london.pm/pipermail/london.pm/Week-of-Mon-20080721/014636.html) from 2008 that suggests that the combination of `-de`, `-Dprevis` and `Duserelocatableinc` has a very specific order that it needs to go in for the call to `Configure`. Specifically that `-de` needs to come last or things go wonky.

Investigating whether or not the order of flags to `Configure` matters and I'll report back once I have an answer

jakirkham commented:
Another fun one like this ( https://github.com/conda/conda-build/issues/865 ). Had to use `install_name_tool` to fix it to full path so `conda-build` would detect. We might need to do similar here.

ericdill commented:
Ah neat, thanks for the example.  IIUC that needs to be done in the perl build?

jakirkham commented:
Not sure. I'm looking into it and some other potential options.

ericdill commented:
Thank you for diving down this rabbit hole with me.

ericdill commented:
@jakirkham What are your thoughts on me just adding `perl-devel` to `yum_requirements.txt` and punting on trying to figure out how to fix the perl package?

ericdill commented:
Certainly no need to apologize @jakirkham .  Thanks for the blessing on using yum_requirements.txt to fix this.  Submitted https://github.com/conda-forge/perl-feedstock/issues/4 and linked it back to this PR

ericdill commented:
@jakirkham This should be ready to go.  Build is successful with `perl-devel` in `yum_requirements.txt` and I've opened an issue against perl-feedstock, as requested.  Thanks for all your help on this

jakirkham commented on recipes/epics-base/build.sh:
```diff
@@ -0,0 +1,40 @@
+
+# tear down
+echo "unset EPICS_BASE" >> $DEACTIVATE
+echo "unset EPICS_HOST_ARCH" >> $DEACTIVATE
```
Could we please break out these [de]activate scripts? It was a little hard to catch what they were doing on a first pass. I'm mainly concerned about readability in the future. So, this is not a blocker per se, but we should clean this up somehow.

tacaswell replied:
Sorry, these things are all my doing.  The first pass at getting this recipe way as a frustrating activity and this was the least-boiler plate way to get this done.

ericdill replied:
The EPICS_BASE variable is defined in build.sh.  How do we get that into the activate script?

jakirkham commented on recipes/epics-base/build.sh:
```diff
@@ -0,0 +1,40 @@
+# clean up after self
+unset ACTIVATE
+unset DEACTIVATE
+unset ETC
```
These will be dumped anyways, but they don't do any harm.

jakirkham commented on recipes/epics-base/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+
+requirements:
+  build:
+    - readline
```
Should be pinned to `6.2*`.

ericdill replied:
If I pin to 6.2*, I pick up the readline from the `defaults` channel since it is 6.2-2 and the one on conda-forge is 6.2-0.  Should conda-forge bump its build number?

jakirkham commented on recipes/epics-base/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+  build:
+    - readline
+  run:
+    - readline
```
As above.

jakirkham commented on recipes/epics-base/meta.yaml:
```diff
@@ -0,0 +1,40 @@
+    - softIoc -h
+about:
+    home: http://www.aps.anl.gov/epics
+    license: Epics Open License
```
Should we be including the license file?

ericdill replied:
How do you include arbitrary files in the conda package?

msarahan replied:
Copy it (manually, ugh) into your recipe folder.  The path here is relative in that folder.

ericdill replied:
Thanks for the quick reply.  I am not quite familiar enough with the directory structure during conda-build to know where to copy to.  `cp {{ environ.SRC_DIR/LICENSE }} <path_to_package?>`

ericdill replied:
@msarahan how do I copy a file into the recipe folder. Do I need to do anything special to get it listed in the `info/files` file?

pelson replied:
Any chance we could change this behaviour @msarahan so that if the license file isn't found in the recipe dir, it looks in the source dir?

pelson commented:
Linux failing:

```
+ unset DEACTIVATE
+ unset ETC
+ cp LICENSE build/
cp: cannot create regular file `build/': Is a directory
Command failed: /bin/bash -x -e /opt/conda/conda-bld/work/base-3.14.12.4/conda_build.sh

./scripts/run_docker_build.sh returned exit code 1
```

ericdill commented:
Thanks for looking at the build logs.  It's failing because I'm not quite sure how to include the license file in the build script.  Any suggestion as to how that can be achieved @pelson @jakirkham ?

jakirkham commented on recipes/epics-base/build.sh:
```diff
@@ -0,0 +1,43 @@
+unset DEACTIVATE
+unset ETC
+
+cp LICENSE build/
```
Please change `build/` to `"${SRC_DIR}"` then this should work.

jakirkham replied:
Actually where is this `LICENSE` now? Is it not in `$SRC_DIR`?

jakirkham commented:
Is this something you wish to continue working on @ericdill ? Or would it be better to close it out and let someone else take over?

ericdill commented:
@licode @tacaswell bump so can work on this if you'd like. Thanks for the reminder @jakirkham
